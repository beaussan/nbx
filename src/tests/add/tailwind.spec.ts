import { filesystem } from 'gluegun';
import * as prompts from 'prompts';
import * as latestVersion from 'latest-version';
import Tailwind from '../../commands/add/tailwind';
import {
  expectFailCli,
  expectFailGitCommits,
  expectGitCommits,
  testCli,
} from '../test-helpers.spec';

jest.setTimeout(90000);

const checkPackgeAndPostcssConfig = async () => {
  const [tailwind, craco] = await Promise.all([
    latestVersion('tailwindcss'),
    latestVersion('@craco/craco'),
  ]);
  const packageJson = filesystem.read('package.json', 'json');

  expect(filesystem.read('src/css/tailwind.css', 'utf8')).toMatchSnapshot();
  expect(filesystem.read('tailwind.config.js', 'utf8')).toBeDefined();
  expect(packageJson).toStrictEqual({
    name: 'sample',
    version: '0.1.0',
    private: true,
    dependencies: {
      'react-scripts': '3.3.1',
      'tailwindcss': tailwind,
    },
    devDependencies: {
      '@craco/craco': craco,
    },
    scripts: {
      start: 'craco start',
      build: 'craco build',
      test: 'craco test',
      eject: 'react-scripts eject',
    },
  });
};

testCli({
  runCommand: (args: string[]) => Tailwind.run(args),
  name: 'nbx add:tailwind',
  defaultArgs: ['--no-spinner'],
  tests: [
    {
      name: 'fail if no package.json found',
      runner: expectFailCli({
        errorMessage:
          'There is no package.json not found in the current folder',
        before: async () => {
          filesystem.remove('package.json');
        },
      }),
    },
    {
      name: 'error if no react is found',
      runner: expectFailGitCommits({
        errorMessage: 'This script support only for now create react apps.',
      }),
    },
    {
      name: 'error if tailwind is found',
      runner: expectFailGitCommits({
        packageJson: {
          dependencies: {
            tailwind: '2.0.0',
          },
        },
        errorMessage: 'Tailwind is already installed in this project.',
      }),
    },
    {
      name: 'error if craco is found',
      runner: expectFailGitCommits({
        packageJson: {
          dependencies: {
            '@craco/craco': '2.0.0',
          },
        },
        errorMessage:
          'Craco is already installed, this may bug so in case, I will stop there.',
      }),
    },
    {
      name: 'work on a react application without commits',
      runner: expectGitCommits({
        packageJson: {
          name: 'sample',
          version: '0.1.0',
          private: true,
          dependencies: {
            'react-scripts': '3.3.1',
          },
          scripts: {
            start: 'react-scripts start',
            build: 'react-scripts build',
            test: 'react-scripts test',
            eject: 'react-scripts eject',
          },
        },
        expectGitLog: [],
        before: async () => {
          prompts.inject([false]);
        },
        checks: checkPackgeAndPostcssConfig,
      }),
    },
    {
      name: 'work on a react application with commits',
      runner: expectGitCommits({
        packageJson: {
          name: 'sample',
          version: '0.1.0',
          private: true,
          dependencies: {
            'react-scripts': '3.3.1',
          },
          scripts: {
            start: 'react-scripts start',
            build: 'react-scripts build',
            test: 'react-scripts test',
            eject: 'react-scripts eject',
          },
        },
        expectGitLog: async () => {
          const [tailwind, craco] = await Promise.all([
            latestVersion('tailwindcss'),
            latestVersion('@craco/craco'),
          ]);

          return [
            ':wrench: add script for tailwind to package.json',
            'M\tpackage.json',
            ':wrench: add tailwind css file',
            'A\tsrc/css/tailwind.css',
            ':wrench: craco postcss config file',
            'A\tcraco.config.js',
            ':wrench: add tailwind config file',
            'A\ttailwind.config.js',
            `:heavy_plus_sign: add tailwindcss@${tailwind} as a dependency`,
            'M\tpackage.json',
            'M\tyarn.lock',
            `:heavy_plus_sign: add @craco/craco@${craco} as a dev dependency`,
            'M\tpackage.json',
            'M\tyarn.lock',
          ];
        },
        before: async () => {
          prompts.inject([true]);
        },
        checks: checkPackgeAndPostcssConfig,
      }),
    },
  ],
});
