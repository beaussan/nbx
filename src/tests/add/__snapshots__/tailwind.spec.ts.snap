// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nbx add:tailwind should error if craco is found 1`] = `Array []`;

exports[`nbx add:tailwind should error if no react is found 1`] = `Array []`;

exports[`nbx add:tailwind should error if tailwind is found 1`] = `Array []`;

exports[`nbx add:tailwind should fail if no package.json found 1`] = `Array []`;

exports[`nbx add:tailwind should print help correctly 1`] = `
Array [
  "add tailwindcss to a project
",
  "USAGE
  $ nbx add:tailwind

OPTIONS
  -h, --help      show CLI help
  -v, --verbose   Verbose output
  --[no-]spinner  Enable spinner in cli output, true by default",
  "",
]
`;

exports[`nbx add:tailwind should work on a react application with commits 1`] = `
"@tailwind base;

/* Write your own custom base styles here */

/* Start purging... */
@tailwind components;
/* Stop purging. */

/* Start purging... */
@tailwind utilities;
/* Stop purging. */

/* Your own custom utilities */
"
`;

exports[`nbx add:tailwind should work on a react application with commits 2`] = `
Array [
  "Adding @craco/craco as a dev dependency",
  "The step was done without any error.",
  "Adding tailwindcss as a dependency",
  "The step was done without any error.",
  "Generating tailwind initial config",
  "The step was done without any error.",
  "Generating craco postcssconfig",
  "The step was done without any error.",
  "Generating tailwind full css",
  "The step was done without any error.",
  "Adding package.json scripts",
  "The step was done without any error.",
]
`;

exports[`nbx add:tailwind should work on a react application without commits 1`] = `
"@tailwind base;

/* Write your own custom base styles here */

/* Start purging... */
@tailwind components;
/* Stop purging. */

/* Start purging... */
@tailwind utilities;
/* Stop purging. */

/* Your own custom utilities */
"
`;

exports[`nbx add:tailwind should work on a react application without commits 2`] = `
Array [
  "Adding @craco/craco as a dev dependency",
  "The step was done without any error.",
  "Adding tailwindcss as a dependency",
  "The step was done without any error.",
  "Generating tailwind initial config",
  "The step was done without any error.",
  "Generating craco postcssconfig",
  "The step was done without any error.",
  "Generating tailwind full css",
  "The step was done without any error.",
  "Adding package.json scripts",
  "The step was done without any error.",
]
`;
